File: c/medium/minimum_cost_path_with_edge_reversals.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

#define MAXN 100000
#define MAXE 200000

typedef struct {
    int to;
    int cost;
} Edge;

typedef struct {
    int node;
    int dist;
} HeapNode;

Edge adj[MAXE];
int head[MAXN], next[MAXE], edge_cnt = 0;

void add_edge(int u, int v, int w) {
    adj[edge_cnt] = (Edge){v, w};
    next[edge_cnt] = head[u];
    head[u] = edge_cnt++;
}

void swap(HeapNode *a, HeapNode *b) {
    HeapNode t = *a;
    *a = *b;
    *b = t;
}

void heap_push(HeapNode *heap, int *size, HeapNode val) {
    heap[(*size)++] = val;
    for (int i = *size - 1; i > 0; ) {
        int p = (i - 1) / 2;
        if (heap[p].dist <= heap[i].dist) break;
        swap(&heap[p], &heap[i]);
        i = p;
    }
}

HeapNode heap_pop(HeapNode *heap, int *size) {
    HeapNode root = heap[0];
    heap[0] = heap[--(*size)];

    for (int i = 0;;) {
        int l = 2*i + 1, r = 2*i + 2, s = i;
        if (l < *size && heap[l].dist < heap[s].dist) s = l;
        if (r < *size && heap[r].dist < heap[s].dist) s = r;
        if (s == i) break;
        swap(&heap[i], &heap[s]);
        i = s;
    }
    return root;
}

int minCost(int n, int edges[][3], int m) {
    for (int i = 0; i < n; i++) head[i] = -1;

    for (int i = 0; i < m; i++) {
        int u = edges[i][0], v = edges[i][1], w = edges[i][2];
        add_edge(u, v, w);
        add_edge(v, u, w << 1);
    }

    int dist[n];
    for (int i = 0; i < n; i++) dist[i] = INT_MAX;
    dist[0] = 0;

    HeapNode heap[MAXE];
    int size = 0;
    heap_push(heap, &size, (HeapNode){0, 0});

    while (size) {
        HeapNode cur = heap_pop(heap, &size);
        if (cur.dist > dist[cur.node]) continue;
        if (cur.node == n - 1) return cur.dist;

        for (int e = head[cur.node]; e != -1; e = next[e]) {
            int nd = cur.dist + adj[e].cost;
            int v = adj[e].to;
            if (nd < dist[v]) {
                dist[v] = nd;
                heap_push(heap, &size, (HeapNode){v, nd});
            }
        }
    }
    return -1;
}

File: rust/medium/minimum_cost_path_with_edge_reversals.rs
use std::cmp::Reverse;
use std::collections::BinaryHeap;

pub fn min_cost(n: usize, edges: Vec<Vec<i32>>) -> i32 {
    let mut adj = vec![Vec::new(); n];

    for e in edges {
        let (u, v, w) = (e[0] as usize, e[1] as usize, e[2]);
        adj[u].push((w, v));
        adj[v].push((w << 1, u));
    }

    let mut dist = vec![i32::MAX; n];
    let mut heap = BinaryHeap::new();

    dist[0] = 0;
    heap.push(Reverse((0, 0)));

    while let Some(Reverse((cost, u))) = heap.pop() {
        if cost > dist[u] { continue; }
        if u == n - 1 { return cost; }

        for &(w, v) in &adj[u] {
            let nc = cost + w;
            if nc < dist[v] {
                dist[v] = nc;
                heap.push(Reverse((nc, v)));
            }
        }
    }
    -1
}
